developerportal/apps/ingestion/utils.py

```diff
@@ -0,0 +1,261 @@
+            )
+        )
+
+    return output
+
+
+def _get_model(model_name):
+    if model_name == "Video":
+        module = video_models
+    elif model_name == "ExternalArticle":
+        module = externalcontent_models
+    else:
+        raise NotImplementedError()
+
+    return getattr(module, model_name)
+
+
+def ingest_content(type_: str):
+    """For the given type_:
+    * fetch the relevant data from the configured sources
+    * for each item of source data:
+        * create an appropriate ExternalContent page subclass, as a draft
+    * update the last_sync timestamp for each configured source
+    * submit all just-created draft pages for moderation
+    """
+
+    _now = tz_now()
+    model_name = type_
+
+    configs = IngestionConfiguration.objects.filter(integration_type=type_)
+
+    Model = _get_model(model_name)
+
+    ingestion_user = User.objects.get(username=INGESTION_USER_USERNAME)
+
+    for config in configs:
+        with transaction.atomic():
+            data_from_source = fetch_external_data(
+                feed_url=config.source_url, last_synced=config.last_sync
+            )
+            config.last_sync = _now
+            config.save()
+
+            for data in data_from_source:
+                data.update(owner=ingestion_user)
+                try:
+                    draft_page = generate_draft_from_external_data(
+                        model=Model, data=data
+                    )
+                except ValidationError as ve:
+                    logger.warning("Problem ingesting article from %s: %s", data, ve)
+                else:
+                    draft_page.owner = ingestion_user
+                    draft_page.save()
+                    # This is a little messy: we save a new revision and ping the
+                    # moderator, but because it's within a transaction if something
+                    # fails we will roll it back at the DB level, but
+                    # the emails notifying Moderators will have been sent for pages
+                    # which will not exist. It's not ideal, but safer than having
+                    # the moderation request happen outside of a transaction, else
+                    # that would risk saved draft-saved pages existing with no
+                    # owner and no 'for moderation' flag set.
+                    revision = draft_page.save_revision(
+                        submitted_for_moderation=True, user=ingestion_user
+                    )
+                    if not send_notification(
+                        page_revision_id=revision.id,
+                        notification="submitted",
+                        excluded_user_id=ingestion_user.id,
+                    ):
+                        logger.warning(
+                            "Failed to send notification that %s was created.",
+                            draft_page,
+                        )
+
+
+def _store_external_image(image_url: str) -> MozImage:
```

Comment:
shouldn't this also be `@transaction.atomic()`?

Reply:
I didn't put it in a transaction because it only touches one model/one DB table, so there's no related-but-not-failed state to roll back. It's called from something that _is_ inside an atomic transaction, though, so if this or something else in that transaction fails, they all get rolled back at the DB level (if not at the file-storage level – which could be nice to do, but not essential I think)